docs/vesting.md

```diff
@@ -0,0 +1,119 @@
+
+
+#### Example usage of vesting tokens
+
+```
+
+contract_instance.vest(
+grantee=ContractAddress,
+vested_timestamp=1650000000
+);
+```
+
+### Parameters
+
+- grantee: The address of the grantee.
+- vested_timestamp: the timestamp at which the tokens vested. This vested_timestamp must be exactly the same as the milestone added.
+
+## Adding a Linear Vesting Schedule
+
+To add a linear vesting schedule, we use the add_linear_vesting_schedule function. This function divides the total amount to be vested into increments and schedules them over specified periods.
+
+#### Example usage of adding a Linear vesting schedule
+
+```
+
+contract_instance.add_linear_vesting_schedule(
+first_vest=1650000000,
+period=2592000, // 30 days in seconds
+increments_count=10,
+total_amount=1200,
+grantee=ContractAddress
+);
+
+```
+
+### Parameters
+
+- first_vest: The timestamp of the first vesting event.
+- period: The time interval between vesting events.
+- increments_count: The number of vesting increments.
+- total_amount: The total amount of tokens to be vested.
+- grantee: The address of the grantee.
+
+#### Cliff Period
+
+A cliff period is the initial period in a vesting schedule during which no tokens are vested. Tokens start to vest only after the cliff period ends. This is useful for ensuring that the grantee remains with the project for a minimum period before receiving any tokens.
+
+##### How Cliff Works
+When using the add_linear_vesting_schedule method, you can set the cliff period by adjusting the first_vest parameter. The first_vest timestamp should be set to a date in the future, representing the end of the cliff period.
+
+* For example, if the current date is January 1, 2024, and you want a 6-month cliff period, the first_vest timestamp should be set to July 1, 2024. Tokens will start vesting only after this date.
+
+## Customization
+
+To customize the Vesting contract for specific protocol’s need, use the methods defined below;
+
+- Custom Proposals: To add new vesting schedules, the component should be part of a custom proposal that calls either add_linear_vesting_schedule or add_vesting_milestone.
+- Contract Integration: The vest function should be exported externally from the contract to allow protocols to execute the vesting process.
+
+The Vesting contract is composed of the following:
+
+- VestingImpl: Implementation of the IVesting trait for managing vesting operations.
+- Storage: Storage component for storing vesting milestones.
+
+### Events
+
+The Vesting contract emits the following events:
+
+- VestingMilestoneAdded: This indicates the addition of a new vesting milestone.
+- Vested: This indicates that tokens have been vested for a grantee.
+
+Note that the methods `add_vesting_milestone ` and `add_linear_vesting_schedule` can only be called by the contract itself to prevent unauthorized modifications.
```

Comment:
Again! What question does this _answer_? How should this be used with Konoha? How can the contract call itself?

Please refer to my first review. You fixed the few concrete issues I raised, but did not address the bigger picture.

Reply:
It's answers the question of the current limitations to it.

> Again! What question does this _answer_? How should this be used with Konoha? How can the contract call itself?
>

Reply:
This will help to avoid Self-call Restrictions.